pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,141 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
+    #
+    # - Each timestamp in the daytime power curve that satisfies the
+    #   clipping criteria[*] is flagged.
+    #
+    # - The clipping threshold is calculated as the mean of the
+    #   longest flagged period in the daytime power curve
+    #
+    # [*] clipping criteria: a timestamp satisfies the clipping
+    # criteria if the derivative of the daytime power curve at that
+    # time is less than `derivative_max` and the value of the daytime
+    # power curve is greater than `power_min` times the median of the
+    # daytime power curve.
+    #
+    # Based on the PVFleets QA Analysis project
+    if not freq:
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds / 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds / 60
+
+    # Use the derivative of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(ac_power)
+    normalized_power = powercurve / powercurve.max()
+    power_derivative = (normalized_power.diff()
+                        / normalized_power.index.to_series().diff()) * freq
+
+    clipped_times = _clipped(powercurve, power_derivative,
+                             powercurve.median() * power_min,
+                             derivative_max)
+    clipping_cumsum = (~clipped_times).cumsum()
+    # get the value of the cumulative sum the longest True span
+    longest_clipped = clipping_cumsum.value_counts().idxmax()
+    # select the longest span that satisfies the clipping criteria
+    longest = powercurve[clipping_cumsum == longest_clipped]
+
+    if longest.index[-1] - longest.index[0] >= 60:
+        # if the period of clipping is at least 60 minutes then we
+        # have enough evidence to determine the clipping threshold.
+        return longest.mean()
+    return None
+
+
+def threshold(ac_power, derivative_max=0.0035,
+              power_quantile=0.995, frequency_quantile=0.25,
+              freq=None):
+    """Detect clipping based on a maximum power threshold.
+
+    A clipping threshold is calculated from the AC power data and any
+    power greater than the threshold is flagged as clipped.
+
+    To calculate the clipping threshold, `ac_power` is aggregated at
+    each minute of the day. Low power data is removed to eliminate
+    night-time periods and the 99.5% quantile is computed at each
+    minute. If the absolute value of the derivative of the 99.5%
+    quantile (with respect to time) is less than `derivative_max` for
+    a continuous period of at least one hour then clipping is
+    indicated. The mean power for that period is used as the
+    threshold. If there are multiple periods with a derivative less
+    than `derivative_max` then the longest period is used to compute
+    the threshold.
+
+    Parameters
+    ----------
+    ac_power : Series
+        DatetimeIndexed series of AC power data.
+    derivative_max : float, default 0.0035
```

Comment:
This needs a statement that it's a fraction of AC power normalized by maximum AC power.

Reply:
I think it is actually the daily power curve (99.5% quantile of `ac_power` at each minute of the day) that is normalized. I'm not sure how to communicate that in the description of the parameter. I think if I spend a bit of time writing a better _overall_ description of the algorithm (and define names like "daily power curve") it will help with describing the parameters.

Reply:
I have rewritten the description of how the clipping threshold is calculated, hopefully in doing so the meaning of all the parameters is also much more clear, but I will need to re-read it after some time working on something else to know for sure...